### PR TITLE
Add filters to music library

### PR DESCRIPTION
**Pull request by 1hitsong**
**Changes**
Add Genre & Year filters to the music library as found in web client.

Doesn't add Year filter when view is set to artists as it will never return data.

### CODE REVIEW

**Comment by cewert** [comment]
I know this doesn't say merge conflict but the required workflow won't run until you merge the latest version of unstable into your PR branch FYI

File: components/ItemGrid/MusicLibraryView.brs [hunk 1]
```diff
@@ -197,7 +206,103 @@ sub loadInitialItems()
...
+sub FilterDataLoaded(msg)
+    options = {}
+    options.filter = []
+    options.favorite = []
+
+    setMusicOptions(options)
+
+    data = msg.GetData()
+    m.getFiltersTask.unobserveField("filters")
+
```
Comment (cewert) [inline components/ItemGrid/MusicLibraryView.brs]:
```suggestion
    m.getFiltersTask.unobserveField("filters")
	m.getFiltersTask.content = []
	
```

Best practice to empty the task after grabbing the data

File: components/ItemGrid/MusicLibraryView.brs [hunk 2]
```diff
@@ -197,7 +206,103 @@ sub loadInitialItems()
...
+        end if
+    end if
+
+    setSelectedOptions(options)
+
+    m.options.options = options
+end sub
+
+' Data to display when options button selected
+sub setSelectedOptions(options)
```
Comment (cewert) [inline components/ItemGrid/MusicLibraryView.brs]:
```suggestion
sub SetSelectedOptions(options)
```

The other functions are all CamelCase

File: components/ItemGrid/MusicLibraryView.brs [hunk 3]
```diff
@@ -197,7 +206,103 @@ sub loadInitialItems()
...
+    options.filter = []
+    options.favorite = []
+
+    setMusicOptions(options)
+
+    data = msg.GetData()
+    m.getFiltersTask.unobserveField("filters")
+    m.getFiltersTask.filters = {}
+
+    if not isValid(data) then return
```
Comment (cewert) [inline components/ItemGrid/MusicLibraryView.brs]:
```suggestion
	data = msg.GetData()
    m.getFiltersTask.unobserveField("filters")
    m.getFiltersTask.filters = {}

    if not isValid(data) then return

    options = {}
    options.filter = []
    options.favorite = []

    setMusicOptions(options)
```

Move validation to top of function unless we need to call `setMusicOptions()` first?

**Review by cewert — CHANGES_REQUESTED** [review]
I had my filter set to favorites then opened the options and selected genres, this changed the filter selected from "favorite" to "all", then after selecting a genre the section changed as expected.

What I'm getting at is I don't think the selection should change to "all after selecting genre. I should be able to change my mind and hit back to restore my previous settings.

**Comment by 1hitsong** [comment]
@cewert This is ready for re-review.

**Review by cewert — CHANGES_REQUESTED** [review]
These might be edge cases but here's what I got (I'm using demo server to test):
- After changing my filter selection from favorites to Genres, if the user changes his mind while the popup is still open and selects a different dropdown selection like "Years" the filter selection reverts to favorites. To reproduce:
  - Open Music library, hit options to open filters, select "favorites", hit options to close filters, hit option to open filters, select "Genres" and then Electronic, before closing the filter popup hit left and down to highlight "Years", select "Years", note how the selection checkmark is now shown by favorites and not "Genres"
- After selecting an option from the filter dropdown/pullout menu thing, if the user changes his filter selection (to favorites for example), whatever the user had ticked on the dropdown/pullout menu stays ticked. To reproduce:
  - Open Music library, hit options to open filters, select "Genre -> Electronic", do not close filter popup, hit left, select favorites, hit down to highlight "Genres" again, note that Electronic is still ticked, select "Genres", Genre now shows a tick and doesn't require me to select Electronic.
